File: dfs/closest_bst_value.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# Approach 1: DFS/Recursive
# @param {TreeNode} root
# @param {Float} target
# @return {Integer}
def closest_value(root, target)
    @closest = Float::INFINITY
    dfs(root, target)
    @closest
end

def dfs(node, trgt)
    return if !node
    
    @closest = node.val if (node.val - trgt).abs < (@closest - trgt).abs
    dfs(node.left, trgt) if trgt < node.val
    dfs(node.right, trgt) if trgt > node.val
end

# Approach 2: Iterative
# @param {TreeNode} root
# @param {Float} target
# @return {Integer}
def closest_value_iter(root, target)
    closest = root.val
    while root
        closest = root.val if (root.val - target).abs < (closest - target).abs
        root = target < root.val ? root.left : root.right
    end

    closest
end


# 270. Closest Binary Search Tree Value
# https://leetcode.com/problems/closest-binary-search-tree-value/description/

# Approach 1: Recurisve
# Time: O(n)
# Space: O(n)

# Approach 2: Iterative
# Time: O(n)
# Space: O(1)


require 'test/unit'
extend Test::Unit::Assertions

# Input: root = [4,2,5,1,3], target = 3.714286
#     4
#    / \
#   2   5
#  / \
# 1   3
root = TreeNode.new(4)
node2a, node2b = TreeNode.new(2), TreeNode.new(5)
root.left, root.right = node2a, node2b
node3a, node3b = TreeNode.new(1), TreeNode.new(3)
node2a.left, node2a.right = node3a, node3b

assert_equal(closest_value(root, 3.714286), 4)
assert_equal(closest_value_iter(root, 3.714286), 4)

File: google/closest_bst_value.py
# Recursive
class Solution:
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        self.closest = float('inf')
        
        def helper(root, value):
            if not root:
                return
            if abs(root.val - target) < abs(self.closest - target):
                self.closest = root.val
                
            # Target should be located on left subtree
            if target < root.val:
                helper(root.left, target)
                
            # target should be located on right subtree
            if target > root.val:
                helper(root.right, target)
        
        helper(root, target)
        return self.closest

# Iterative
class Solution(object):
    def closestValue(self, root, target):
        closest = root.val
        while root:
            if abs(root.val - target) < abs(closest - target):
                closest = root.val
            root = root.left if target < root.val else root.right
        return closest

# > Time Complexity O(N)
# > Space Complexity O(1)
